File: src/Sdks/SdkHelper.fs
module Helium.Sdks.SdkHelper

open System


exception UnknownPlatformException

let private (|IgnoreCase|_|) (s1: string) (s2: string) =
    if String.Compare(s1, s2, StringComparison.InvariantCultureIgnoreCase) = 0 then
        Some()
    else
        None

let parseOperatingSystem: string -> SdkOperatingSystem = function
    | IgnoreCase "linux" -> SdkOperatingSystem.Linux
    | IgnoreCase "windows" | IgnoreCase "win" | IgnoreCase "win32" -> SdkOperatingSystem.Windows
    | _ -> raise UnknownPlatformException
    
    
let parseArch: string -> SdkArch = function
    | IgnoreCase "x64" | IgnoreCase "x86_64" | IgnoreCase "amd64" -> SdkArch.Amd64
    | IgnoreCase "x86" | IgnoreCase "x32" -> SdkArch.X86
    | IgnoreCase "arm" | IgnoreCase "arm32" -> SdkArch.Arm
    | IgnoreCase "arm64" | IgnoreCase "aarch64" -> SdkArch.Aarch64
    | IgnoreCase "ppc64le" -> SdkArch.Ppc64le
    | IgnoreCase "s390x" -> SdkArch.S390x
    | _ -> raise UnknownPlatformException
    

File: src/Sdks/SdkHelper.cs
using System;

namespace Helium.Sdks
{
    public static class SdkHelper
    {
        public static SdkOperatingSystem ParseOperatingSystem(string os) {

            bool Check(string name) => os.Equals(name, StringComparison.InvariantCultureIgnoreCase);
            
            
            if(Check("linux")) return SdkOperatingSystem.Linux;
            else if(Check("windows") || Check("win") || Check("win32")) return SdkOperatingSystem.Windows;
            else throw new UnknownPlatformException();
        }

        public static SdkArch ParseArch(string arch) {

            bool Check(string name) => arch.Equals(name, StringComparison.InvariantCultureIgnoreCase);
            
            
                if(Check("x64") || Check("x86_64") || Check("amd64")) return SdkArch.Amd64;
                else if(Check("x86") || Check("x32")) return SdkArch.X86;
                else if(Check("arm") || Check("arm32")) return SdkArch.Arm;
                else if(Check("arm64") || Check("aarch64")) return SdkArch.Aarch64;
                else if(Check("ppc64le")) return SdkArch.Ppc64le;
                else if(Check("s390x")) return SdkArch.S390x;
                else throw new UnknownPlatformException();
            
        }
    }
}